### DISCUSSION TITLE
Alternative 2D/3D subsystems and WASM

### DISCUSSION THREAD

**Discussion by Loosetooth** [jscad/OpenJSCAD.org#1265] [General]
A while back I noticed that the offset API of JSCAD does not work very well. See #1220 for example. I had a look through the code, and it looks like the 2D system from JSCAD is really prototype code compared to others, like clipperJS/clipper/clipper2.

As an example, the [manifold](https://github.com/elalish/manifold) library exposes clipper2 through their API. And it seems to work really well.

The question: would it be possible for JSCAD to also integrate clipper2 for example as a 2D subsystem? Or does this go against the 'pure javascript' idea of JSCAD? (As this would require c++ code running via webassembly.)

Following this train of thought: perhaps it would be also possible to re-create the JSCAD API, but using manifold as the 'backend'. This would of course be a separate project, but it would allow us to use the clipper2 subsystem out of the box. It would also probably run a lot faster.

**Comment by hrgdavor** [comment]
I ma closely following manifold library, I am subscriber to the emails on git there. My idea was at some point to somehow make using manifold as optional replacement for csg.js.

I am very impressed by manifold library and the development of it, but I think a transitional step would be nicer, so if you have a script that worked with csg.js, and fails with manifold, you can revert temporarily until issue is resolved, or workaround found.

However, my work related to jscad is going very slowly for one year now. I would gladly tackle this task if I find some time. Until then, I provide my thoughts on the topic and could help if sbdy else would attempt it.
